Python wheel `install.sh` is using deprecated instructions as of setuptools v0.60.0

## Problem

Hi. :wave: With the release of [`setuptools` `v0.60.0` on 2021-12-19](https://pypi.org/project/setuptools/60.0.0/), the Python [wheel `install.sh`](https://github.com/xrootd/xrootd/blob/af217b9a5c196f921bdc460930390828a9f927f4/packaging/wheel/install.sh) script will produce warnings (here shown for Python 3.9) of the form 

```pytb
/usr/local/venv/lib/python3.9/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
  warnings.warn(
/usr/local/venv/lib/python3.9/site-packages/setuptools/command/easy_install.py:156: EasyInstallDeprecationWarning: easy_install command is deprecated. Use build and pip and other standards-based tools.
  warnings.warn(
```

and the resulting directory structure in `site-packages` will change from installing `pyxrootd` in `site-packages`'s top level (correctly)

```console
root@d68489115a42:/# ls -lhtr /usr/local/venv/lib/python3.9/site-packages/
total 48K
drwxr-xr-x 2 root root 4.0K Dec 21 05:23 wheel-0.37.0.dist-info
drwxr-xr-x 5 root root 4.0K Dec 21 05:23 wheel
drwxr-xr-x 5 root root 4.0K Dec 21 05:23 pip
drwxr-xr-x 2 root root 4.0K Dec 21 05:23 pip-21.3.1.dist-info
drwxr-xr-x 7 root root 4.0K Dec 21 06:22 setuptools
drwxr-xr-x 6 root root 4.0K Dec 21 06:22 pkg_resources
-rw-r--r-- 1 root root  152 Dec 21 06:22 distutils-precedence.pth
drwxr-xr-x 3 root root 4.0K Dec 21 06:22 _distutils_hack
drwxr-xr-x 2 root root 4.0K Dec 21 06:22 setuptools-59.8.0.dist-info
-rw-r--r-- 1 root root  252 Dec 21 06:24 xrootd-v5.4.0-py3.9.egg-info
drwxr-xr-x 3 root root 4.0K Dec 21 06:24 pyxrootd
drwxr-xr-x 4 root root 4.0K Dec 21 06:24 XRootD
```

to installing `pyxrootd` below in `site-packages/xrootd-5.4.0-py3.9-linux-x86_64.egg`,

```console
root@ec762b69abe0:/# ls -lhtr /usr/local/venv/lib/python3.9/site-packages/
total 44K
drwxr-xr-x 2 root root 4.0K Dec 21 05:23 wheel-0.37.0.dist-info
drwxr-xr-x 5 root root 4.0K Dec 21 05:23 wheel
drwxr-xr-x 5 root root 4.0K Dec 21 05:23 pip
drwxr-xr-x 2 root root 4.0K Dec 21 05:23 pip-21.3.1.dist-info
drwxr-xr-x 2 root root 4.0K Dec 21 06:24 setuptools-60.0.0.dist-info
drwxr-xr-x 7 root root 4.0K Dec 21 06:24 setuptools
drwxr-xr-x 6 root root 4.0K Dec 21 06:24 pkg_resources
-rw-r--r-- 1 root root  151 Dec 21 06:24 distutils-precedence.pth
drwxr-xr-x 3 root root 4.0K Dec 21 06:24 _distutils_hack
drwxr-xr-x 5 root root 4.0K Dec 21 06:26 xrootd-5.4.0-py3.9-linux-x86_64.egg
-rw-r--r-- 1 root root   38 Dec 21 06:26 easy-install.pth
root@ec762b69abe0:/# ls -lhtr /usr/local/venv/lib/python3.9/site-packages/xrootd-5.4.0-py3.9-linux-x86_64.egg/
total 12K
drwxr-xr-x 3 root root 4.0K Dec 21 06:26 pyxrootd
drwxr-xr-x 4 root root 4.0K Dec 21 06:26 XRootD
drwxr-xr-x 2 root root 4.0K Dec 21 06:26 EGG-INFO
```

which is atypical and not good in general as you're now dependent on the deprecated `easy-install.pth`.

## Minimal Reproducible Example

To make this easier to see, [here's a GitHub Gist](https://gist.github.com/matthewfeickert/6ff43901c7a76e07b5b92620be1496c5) with a fully reproducible example in the form of a Dockerfile and a Makefile that will produce a working (`setuptools` `v59.8.0`) and failing (`setuptools` `v60.0.0`) example. c.f. `setuptools-60.0.0-fail.log.txt` for the warnings I describe above. (I know that all the packages I have in `packages.txt` aren't needed, but that was easiest for me to grab from an existing build, so feel free to remove the unneeded ones if you want.)

## Requested Action

Can you please update the install script to be compatible with modern `setuptools`? This will help a lot when people are trying to install xrootd with its full Python bindings in scenarios that require some "creative" installs (I have [a few of them](https://gitlab.cern.ch/atlas-amglab/atlstats/-/commit/69b2fbe40d5fd46ae56d0ef2a9b0a911d4421985), sadly).

using master 30c089925 i get an error when building the wheel:
```
 Building wheel for xrootd (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/bin/python3 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/home/adrian/GIT-REPOS/xrootd/xrootdbuild/bindings/python/setup.py'"'"'; __file__='"'"'/home/adrian/GIT-REPOS/xrootd/xrootdbuild/bindings/python/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /home/adrian/tmp/pip-wheel-awqsi185
       cwd: /home/adrian/GIT-REPOS/xrootd/xrootdbuild/bindings/python/
  Complete output (63 lines):
  XRootD library dir:     /home/adrian/work-GRID/xrootd/xrootdbuild/src
  XRootD src include dir: /home/adrian/work-GRID/xrootd/src
  XRootD bin include dir: /home/adrian/work-GRID/xrootd/xrootdbuild/src
  Version:                v20220106-30c089925
  /home/adrian/.local/lib/python3.10/site-packages/setuptools/dist.py:501: UserWarning: The version specified ('v20220106-30c089925') is an invalid version, this may not work as expected with newer versions of setuptools, pip, and PyPI. Please see PEP 440 for more details.
    warnings.warn(
  running bdist_wheel
  running build
  running build_py
  running build_ext
  /home/adrian/.local/lib/python3.10/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
    warnings.warn(
  installing to build/bdist.linux-x86_64/wheel
  running install
  running install_lib
  creating build/bdist.linux-x86_64
  creating build/bdist.linux-x86_64/wheel
  creating build/bdist.linux-x86_64/wheel/pyxrootd
  copying build/lib.linux-x86_64-3.10/pyxrootd/client.cpython-310-x86_64-linux-gnu.so -> build/bdist.linux-x86_64/wheel/pyxrootd
  copying build/lib.linux-x86_64-3.10/pyxrootd/__init__.py -> build/bdist.linux-x86_64/wheel/pyxrootd
  creating build/bdist.linux-x86_64/wheel/XRootD
  creating build/bdist.linux-x86_64/wheel/XRootD/client
  copying build/lib.linux-x86_64-3.10/XRootD/client/filesystem.py -> build/bdist.linux-x86_64/wheel/XRootD/client
  copying build/lib.linux-x86_64-3.10/XRootD/client/responses.py -> build/bdist.linux-x86_64/wheel/XRootD/client
  copying build/lib.linux-x86_64-3.10/XRootD/client/copyprocess.py -> build/bdist.linux-x86_64/wheel/XRootD/client
  copying build/lib.linux-x86_64-3.10/XRootD/client/glob_funcs.py -> build/bdist.linux-x86_64/wheel/XRootD/client
  copying build/lib.linux-x86_64-3.10/XRootD/client/__init__.py -> build/bdist.linux-x86_64/wheel/XRootD/client
  copying build/lib.linux-x86_64-3.10/XRootD/client/env.py -> build/bdist.linux-x86_64/wheel/XRootD/client
  copying build/lib.linux-x86_64-3.10/XRootD/client/file.py -> build/bdist.linux-x86_64/wheel/XRootD/client
  copying build/lib.linux-x86_64-3.10/XRootD/client/_version.py -> build/bdist.linux-x86_64/wheel/XRootD/client
  copying build/lib.linux-x86_64-3.10/XRootD/client/url.py -> build/bdist.linux-x86_64/wheel/XRootD/client
  copying build/lib.linux-x86_64-3.10/XRootD/client/finalize.py -> build/bdist.linux-x86_64/wheel/XRootD/client
  copying build/lib.linux-x86_64-3.10/XRootD/client/flags.py -> build/bdist.linux-x86_64/wheel/XRootD/client
  copying build/lib.linux-x86_64-3.10/XRootD/client/utils.py -> build/bdist.linux-x86_64/wheel/XRootD/client
  copying build/lib.linux-x86_64-3.10/XRootD/__init__.py -> build/bdist.linux-x86_64/wheel/XRootD
  running install_egg_info
  /home/adrian/.local/lib/python3.10/site-packages/pkg_resources/__init__.py:116: PkgResourcesDeprecationWarning: v20220106-30c089925 is an invalid version and will not be supported in a future release
    warnings.warn(
  running egg_info
  creating xrootd.egg-info
  writing xrootd.egg-info/PKG-INFO
  writing dependency_links to xrootd.egg-info/dependency_links.txt
  writing top-level names to xrootd.egg-info/top_level.txt
  writing manifest file 'xrootd.egg-info/SOURCES.txt'
  reading manifest file 'xrootd.egg-info/SOURCES.txt'
  writing manifest file 'xrootd.egg-info/SOURCES.txt'
  Copying xrootd.egg-info to build/bdist.linux-x86_64/wheel/xrootd-v20220106_30c089925-py3.10.egg-info
  running install_scripts
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/home/adrian/GIT-REPOS/xrootd/xrootdbuild/bindings/python/setup.py", line 66, in <module>
      setup( name             = 'xrootd',
    File "/usr/lib64/python3.10/distutils/core.py", line 148, in setup
      dist.run_commands()
    File "/usr/lib64/python3.10/distutils/dist.py", line 966, in run_commands
      self.run_command(cmd)
    File "/usr/lib64/python3.10/distutils/dist.py", line 985, in run_command
      cmd_obj.run()
    File "/home/adrian/.local/lib/python3.10/site-packages/wheel/bdist_wheel.py", line 328, in run
      impl_tag, abi_tag, plat_tag = self.get_tag()
    File "/home/adrian/.local/lib/python3.10/site-packages/wheel/bdist_wheel.py", line 278, in get_tag
      assert tag in supported_tags, "would build wheel with unsupported tag {}".format(tag)
  AssertionError: would build wheel with unsupported tag ('cp310', 'cp310', 'linux_x86_64')
  ----------------------------------------
  ERROR: Failed building wheel for xrootd
  Running setup.py clean for xrootd
Failed to build xrootd
```

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -63,6 +63,7 @@ jobs:
             -DCMAKE_INSTALL_PREFIX=/usr/local/ \
             -DPYTHON_EXECUTABLE=$(command -v python3) \
             -DENABLE_TESTS=ON \
+            -DPIP_VERBOSE=ON \
             -S xrootd \
             -B build
         cmake3 build -LH
@@ -135,6 +136,7 @@ jobs:
             -DCMAKE_INSTALL_PREFIX=/usr/local/ \
             -DPYTHON_EXECUTABLE=$(command -v python3) \
             -DENABLE_TESTS=ON \
+            -DPIP_VERBOSE=ON \
             -S xrootd \
             -B build
         cmake3 build -LH
@@ -202,6 +204,7 @@ jobs:
             -DCMAKE_INSTALL_PREFIX=/usr/ \
             -DPYTHON_EXECUTABLE=$(command -v python2) \
             -DENABLE_TESTS=ON \
+            -DPIP_VERBOSE=ON \
             -S xrootd \
             -B build
         cmake3 build -LH

--- a/bindings/python/CMakeLists.txt
+++ b/bindings/python/CMakeLists.txt
@@ -32,6 +32,12 @@ endif()
 
 configure_file(${SETUP_PY_IN} ${SETUP_PY})
 
+if( PIP_VERBOSE )
+  set(PIP_INSTALL_VERBOSE_FLAG "--verbose")
+else()
+  set(PIP_INSTALL_VERBOSE_FLAG "")
+endif()
+
 # Check it the Python interpreter has a valid version of pip
 execute_process(
         COMMAND ${PYTHON_EXECUTABLE} -m pip --version
@@ -53,7 +59,7 @@ if ( NOT ${VALID_PIP_EXIT_CODE} EQUAL 0 )
 
   # https://docs.python.org/3/install/#splitting-the-job-up
   add_custom_command(OUTPUT ${OUTPUT}
-                     COMMAND ${PYTHON_EXECUTABLE} ${SETUP_PY} build
+                     COMMAND ${PYTHON_EXECUTABLE} ${SETUP_PY} ${PIP_INSTALL_VERBOSE_FLAG} build
                      DEPENDS ${DEPS})
 
   add_custom_target(python_target ALL DEPENDS ${OUTPUT} XrdCl)
@@ -73,6 +79,7 @@ if ( NOT ${VALID_PIP_EXIT_CODE} EQUAL 0 )
     CODE
     "EXECUTE_PROCESS(
       COMMAND ${PYTHON_EXECUTABLE} ${SETUP_PY} install \
+                ${PIP_INSTALL_VERBOSE_FLAG} \
                 --prefix \$ENV{DESTDIR}/${CMAKE_INSTALL_PREFIX} \
                 ${DEB_INSTALL_ARGS} \
                 --record PYTHON_INSTALLED
@@ -85,6 +92,7 @@ else()
     CODE
     "EXECUTE_PROCESS(
       COMMAND ${PYTHON_EXECUTABLE} -m pip install \
+                ${PIP_INSTALL_VERBOSE_FLAG} \
                 --force-reinstall \
                 --prefix \$ENV{DESTDIR}/${CMAKE_INSTALL_PREFIX} \
                 ${CMAKE_CURRENT_BINARY_DIR}

--- a/bindings/python/CMakeLists.txt
+++ b/bindings/python/CMakeLists.txt
@@ -32,25 +32,62 @@ endif()
 
 configure_file(${SETUP_PY_IN} ${SETUP_PY})
 
-add_custom_command(OUTPUT ${OUTPUT}
-                   COMMAND ${PYTHON_EXECUTABLE} ${SETUP_PY} build
-                   DEPENDS ${DEPS})
+# Check it the Python interpreter has a valid version of pip
+execute_process(
+        COMMAND ${PYTHON_EXECUTABLE} -m pip --version
+        RESULT_VARIABLE VALID_PIP_EXIT_CODE
+        OUTPUT_QUIET
+)
 
-add_custom_target(python_target ALL DEPENDS ${OUTPUT} XrdCl)
+if ( NOT ${VALID_PIP_EXIT_CODE} EQUAL 0 )
+  # Attempt to still install with deprecated invocation of setup.py
+  message(WARNING
+         " ${PYTHON_EXECUTABLE} does not have a valid pip associated with it."
+         " It is recommended that you install a version of pip to install Python"
+         " packages and bindings. If you are unable to install a version of pip"
+         " through a package manager or with your Python build try using the PyPA's"
+         " get-pip.py bootstrapping script ( https://github.com/pypa/get-pip ).\n"
+         " The installation of the Python bindings will attempt to continue using"
+         " the deprecated method of `${PYTHON_EXECUTABLE} setup.py install`."
+         )
 
-# Get the distribution name on Debian families
-execute_process( COMMAND grep -i ^NAME=\" /etc/os-release
-                 OUTPUT_VARIABLE DEB_DISTRO )
-STRING(REGEX REPLACE "^NAME=\"" "" DEB_DISTRO "${DEB_DISTRO}")
-STRING(REGEX REPLACE "\".*"     "" DEB_DISTRO "${DEB_DISTRO}")
+  # https://docs.python.org/3/install/#splitting-the-job-up
+  add_custom_command(OUTPUT ${OUTPUT}
+                     COMMAND ${PYTHON_EXECUTABLE} ${SETUP_PY} build
+                     DEPENDS ${DEPS})
 
-if( DEB_DISTRO STREQUAL "Debian" OR DEB_DISTRO STREQUAL "Ubuntu" )
-  set(PYTHON_LAYOUT  "unix" CACHE STRING "Python installation layout (deb or unix)")
-  set(DEB_INSTALL_ARGS "--install-layout ${PYTHON_LAYOUT}")
+  add_custom_target(python_target ALL DEPENDS ${OUTPUT} XrdCl)
+
+  # Get the distribution name on Debian families
+  execute_process( COMMAND grep -i ^NAME= /etc/os-release
+                   OUTPUT_VARIABLE DEB_DISTRO )
+  STRING(REGEX REPLACE "^NAME=\"" "" DEB_DISTRO "${DEB_DISTRO}")
+  STRING(REGEX REPLACE "\".*"     "" DEB_DISTRO "${DEB_DISTRO}")
+
+  if( DEB_DISTRO STREQUAL "Debian" OR DEB_DISTRO STREQUAL "Ubuntu" )
+    set(PYTHON_LAYOUT  "unix" CACHE STRING "Python installation layout (deb or unix)")
+    set(DEB_INSTALL_ARGS "--install-layout ${PYTHON_LAYOUT}")
+  endif()
+
+  install(
+    CODE
+    "EXECUTE_PROCESS(
+      COMMAND ${PYTHON_EXECUTABLE} ${SETUP_PY} install \
+                --prefix \$ENV{DESTDIR}/${CMAKE_INSTALL_PREFIX} \
+                ${DEB_INSTALL_ARGS} \
+                --record PYTHON_INSTALLED
+    )"
+  )
+
+else()
+  # Use --force-reinstall to ensure a clean install if a rebuild needs to happen
+  install(
+    CODE
+    "EXECUTE_PROCESS(
+      COMMAND ${PYTHON_EXECUTABLE} -m pip install \
+                --force-reinstall \
+                --prefix \$ENV{DESTDIR}/${CMAKE_INSTALL_PREFIX} \
+                ${CMAKE_CURRENT_BINARY_DIR}
+      )"
+  )
 endif()
-
-install(
-  CODE
-  "EXECUTE_PROCESS(
-    COMMAND ${PYTHON_EXECUTABLE} ${SETUP_PY} install --prefix \$ENV{DESTDIR}/${CMAKE_INSTALL_PREFIX} ${DEB_INSTALL_ARGS} --record PYTHON_INSTALLED)" )
-

--- a/cmake/XRootDDefaults.cmake
+++ b/cmake/XRootDDefaults.cmake
@@ -18,6 +18,7 @@ option( ENABLE_XRDCL     "Enable XRootD client."                                
 option( ENABLE_TESTS     "Enable unit tests."                                             FALSE )
 option( ENABLE_HTTP      "Enable HTTP component."                                         TRUE )
 option( ENABLE_PYTHON    "Enable python bindings."                                        TRUE )
+option( PIP_VERBOSE      "Turn on --verbose flag for pip during Python bindings install." FALSE )
 option( XRDCL_ONLY       "Build only the client and necessary dependencies"               FALSE )
 option( XRDCL_LIB_ONLY   "Build only the client libraries and necessary dependencies"     FALSE )
 option( PYPI_BUILD       "The project is being built for PyPI release"                    FALSE )

--- a/packaging/rhel/xrootd.spec.in
+++ b/packaging/rhel/xrootd.spec.in
@@ -106,6 +106,7 @@ BuildRequires: libmacaroons-devel
 BuildRequires: json-c-devel
 
 %if %{python2only}
+BuildRequires: python2-pip
 BuildRequires: python2-devel
 BuildRequires: python2-setuptools
 %endif

